piptools/scripts/compile.py

```diff
@@ -500,6 +507,25 @@ def cli(
 
     log.debug("")
 
+    # Determine linesep for OutputWriter to use
+    if newline == "preserve":
+        for fname in (output_file.name, *src_files):
+            if os.path.exists(fname):
+                with open(fname, "rb") as existing_file:
+                    existing_text = existing_file.read().decode()
```

Comment:
this keeps a reference to the existing_text for the duration of the program - which might be quite large

Reply:
```python
def determine_linesep(newline, output_file, src_files):
    # Determine linesep for OutputWriter to use
    if newline == "preserve":
        for fname in (output_file.name, *src_files):
            try:
                with open(fname, "rb") as existing_file:
                    existing_text = existing_file.read().decode()
            except FileNotFoundError:
                break
            else:
                if "\r\n" in existing_text:
                    return "\r\n"
                if "\n" in existing_text:
                    return "\n"
    return {
        "native": os.linesep,
        "LF": "\n",
        "CRLF": "\r\n",
        "preserve": "\n",
    }[newline]
```